### DISCUSSION TITLE
关于Agent的设计

### DISCUSSION THREAD

**Discussion by sampsonye** [go-to-cloud/go-to-cloud#8] [Ideas]
## Agent定位
1. Agent是一个K8S集群的**抽象**代表，每一个成功部署的Agent，都属于GTC的托管K8S集群
2. Agent负责接收命令，并将命令的执行结果返回给服务端
3. Agent有自我更新的能力

## 能力实现
1. 使用gRPC流进行长连接
2. 心跳机制确保双端活跃
3. Agent端除k8sconfig外不保存任何数据
4. Agent全生命周期无状态

## 议论点
1. Agent端重启后如何延续之前的命令执行？
2. Server端重启后如何延续接收之前的数据？
3. 如何更新Agent？
4. Agent端提供哪些基础接口？

**Comment by fanhousanbu** [comment]
> ## Agent定位
> 1. Agent是一个K8S集群的**抽象**代表，每一个成功部署的Agent，都属于GTC的托管K8S集群
> 2. Agent负责接收命令，并将命令的执行结果返回给服务端
> 3. Agent有自我更新的能力
> 
> ## 能力实现
> 1. 使用gRPC流进行长连接
> 2. 心跳机制确保双端活跃
> 3. Agent端除k8sconfig外不保存任何数据
> 4. Agent全生命周期无状态
> 
> ## 议论点
> 1. Agent端重启后如何延续之前的命令执行？
> 2. Server端重启后如何延续接收之前的数据？
> 3. 如何更新Agent？
> 4. Agent端提供哪些基础接口？

PRE:
1. Agent当前目标是基于K8s，但未来也会考虑支持裸机部署、app（基于macos的agent)、sdk构建——这些目前不考虑是基于当前的项目定位

2. 更新应该由服务器发起吧？

讨论点：
1. 2. Agent端重启后如何延续之前的命令执行：Agent不负责，我们认为每个命令都是原子的，所以如果失败，则由服务器重发
4: 接收指令，实时反馈执行命令的控制台输出

**Reply by sampsonye** [reply]
1. 如果考虑多平台（未来）那么此刻我们是否需要考虑抽象出一套平台无关的API？
2. 更新由服务端发起，双端需要知道各自的版本，这部分需要思考如何实现
3. Agent命令原子性是个很棒的方向，牛逼~~
